llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp

```diff
@@ -58,25 +94,258 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+  if (ShaderFlags != 0) {
+    SmallVector<Metadata *> MDVals;
+    // FIXME: ShaderFlagsAnalysis pass needs to collect and provide
+    // ShaderFlags for each entry function. Currently, ShaderFlags value
+    // provided by ShaderFlagsAnalysis pass is created by walking *all* the
+    // function instructions of the module. Is it is correct to use this value
+    // for metadata of the empty library entry?
+    MDVals.append(getTagValueAsMetadata(ShaderFlagsTag, ShaderFlags, Ctx));
+    Properties = MDNode::get(Ctx, MDVals);
+  }
+  // Library has an entry metadata with resource table metadata and all other
+  // MDNodes as null.
+  return constructEntryMetadata(nullptr, nullptr, RMD, Properties, Ctx);
 }
 
 static void translateMetadata(Module &M, const DXILResourceMap &DRM,
-                              const dxil::Resources &MDResources,
-                              const ComputedShaderFlags &ShaderFlags) {
-  dxil::ValidatorVersionMD ValVerMD(M);
-  if (ValVerMD.isEmpty())
-    ValVerMD.update(VersionTuple(1, 0));
-  dxil::createShaderModelMD(M);
-  dxil::createDXILVersionMD(M);
+                              const Resources &MDResources,
+                              const ComputedShaderFlags &ShaderFlags,
+                              const ModuleMetadataInfo &MMDI) {
+  LLVMContext &Ctx = M.getContext();
+  IRBuilder<> IRB(Ctx);
+  SmallVector<MDNode *> EntryFnMDNodes;
+
+  emitValidatorVersionMD(M, MMDI);
+  emitShaderModelVersionMD(M, MMDI);
+  emitDXILVersionTupleMD(M, MMDI);
+  NamedMDNode *NamedResourceMD = emitResourceMetadata(M, DRM, MDResources);
+  auto *ResourceMD =
+      (NamedResourceMD != nullptr) ? NamedResourceMD->getOperand(0) : nullptr;
+  // FIXME: Add support to construct Signatures
+  // See https://github.com/llvm/llvm-project/issues/57928
+  MDTuple *Signatures = nullptr;
 
-  emitResourceMetadata(M, DRM, MDResources);
+  if (MMDI.ShaderProfile == Triple::EnvironmentType::Library)
+    EntryFnMDNodes.emplace_back(
+        emitTopLevelLibraryNode(M, ResourceMD, ShaderFlags));
+  else if (MMDI.EntryPropertyVec.size() > 1) {
+    M.getContext().diagnose(DiagnosticInfoModuleFormat(
+        M, "Non-library shader: One and only one entry expected"));
+  }
+
+  for (const EntryProperties &EntryProp : MMDI.EntryPropertyVec) {
+    // FIXME: ShaderFlagsAnalysis pass needs to collect and provide
+    // ShaderFlags for each entry function. For now, assume shader flags value
+    // of entry functions being compiled for lib_* shader profile viz.,
+    // EntryPro.Entry is 0.
+    uint64_t EntryShaderFlags =
+        (MMDI.ShaderProfile == Triple::EnvironmentType::Library) ? 0
+                                                                 : ShaderFlags;
+    if (MMDI.ShaderProfile != Triple::EnvironmentType::Library) {
+      if (EntryProp.ShaderStage != MMDI.ShaderProfile) {
+        M.getContext().diagnose(DiagnosticInfoModuleFormat(
+            M, "Non-library shader: Stage of Shader entry different from "
```

Comment:
"Stage of Shader entry" - the code and diagnostics seem to be inconsistent about when "shader profile" and "shader stage" are used.

Reply:
> "Stage of Shader entry" - the code and diagnostics seem to be inconsistent about when "shader profile" and "shader stage" are used.

"Shader Profile" is expected to convey the target shader profile while "Shader Stage" is associated with the entry. Modified the message to "Non-library shader: Stage of shader entry different from shader target profile"

Reply:
> "Non-library shader: Stage of shader entry different from shader target profile"

So, I think: if this diagnostic is emitted for IR generated by Clang then this is a bug in Clang.  Does that sound right?  The only way to really hit this is for some other frontend to do the wrong thing?

Given that the case we can probably frame the diagnostic using llvm terms like "environment" and "triple" if that helps at all?

Reply:
> > "Non-library shader: Stage of shader entry different from shader target profile"
> 
> So, I think: if this diagnostic is emitted for IR generated by Clang then this is a bug in Clang. Does that sound right? The only way to really hit this is for some other frontend to do the wrong thing?
> 
> Given that the case we can probably frame the diagnostic using llvm terms like "environment" and "triple" if that helps at all?

How about - 
```
Shader stage '<entry-shader-stage>' of entry '<entry>' different from specified target profile '<target-shader-profile>'
```
 where the exact mismatching component is spelt out? An instance of such a message would look like  

```
Shader stage 'cs' of entry 'cs_entry' different from specified target profile 'pixel'
```